File: frontend/src/types/job.ts
// Job module TypeScript types — mirrors backend schemas/job.py

export type JobStatus = "draft" | "active" | "paused" | "closed";
export type JobType =
  | "full_time"
  | "part_time"
  | "contract"
  | "internship"
  | "remote";
export type QuestionType = "text" | "yes_no" | "multiple_choice" | "rating";

// ── Skills ──────────────────────────────────────────────────────────────────

export interface JobSkill {
  id: string;
  skill_name: string;
  is_required: boolean;
}

export interface JobSkillCreate {
  skill_name: string;
  is_required?: boolean;
}

// ── Questions ───────────────────────────────────────────────────────────────

export interface JobQuestion {
  id: string;
  question_text: string;
  question_type: QuestionType;
  options: string[] | null;
  is_required: boolean;
  display_order: number;
}

export interface JobQuestionCreate {
  question_text: string;
  question_type?: QuestionType;
  options?: string[] | null;
  is_required?: boolean;
  display_order?: number;
}

export interface JobQuestionUpdate {
  question_text?: string;
  question_type?: QuestionType;
  options?: string[] | null;
  is_required?: boolean;
  display_order?: number;
}

// ── Pipeline ─────────────────────────────────────────────────────────────────

export interface PipelineStage {
  id: string;
  stage_name: string;
  stage_order: number;
  color: string;
  is_default: boolean;
}

export interface PipelineStageCreate {
  stage_name: string;
  color?: string;
}

export interface PipelineStageUpdate {
  stage_name?: string;
  color?: string;
}

export interface PipelineStageReorderItem {
  id: string;
  stage_order: number;
}

// ── Job CRUD ─────────────────────────────────────────────────────────────────

export interface JobCreate {
  title: string;
  description?: string;
  requirements?: string;
  location?: string;
  job_type?: JobType;
  department?: string;
  salary_min?: number;
  salary_max?: number;
  currency?: string;
  experience_years_min?: number;
  experience_years_max?: number;
  deadline?: string; // ISO datetime string
}

export interface JobUpdate extends Partial<JobCreate> {}

export interface Job {
  id: string;
  title: string;
  description: string | null;
  requirements: string | null;
  location: string | null;
  job_type: JobType;
  department: string | null;
  salary_min: number | null;
  salary_max: number | null;
  currency: string;
  experience_years_min: number | null;
  experience_years_max: number | null;
  status: JobStatus;
  posted_by: string | null;
  company_id: string | null;
  published_at: string | null;
  closed_at: string | null;
  archived_at: string | null;
  deadline: string | null;
  views_count: number;
  applications_count: number;
  created_at: string;
  updated_at: string;
  skills: JobSkill[];
  questions: JobQuestion[];
  pipeline_stages: PipelineStage[];
}

export interface JobListItem {
  id: string;
  title: string;
  location: string | null;
  job_type: JobType;
  department: string | null;
  status: JobStatus;
  views_count: number;
  applications_count: number;
  created_at: string;
  published_at: string | null;
  deadline: string | null;
  skills: JobSkill[];
}

export interface JobListResponse {
  items: JobListItem[];
  total: number;
  page: number;
  page_size: number;
}

// ── Analytics ────────────────────────────────────────────────────────────────

export interface JobCountByStatus {
  draft: number;
  active: number;
  paused: number;
  closed: number;
}

export interface JobCountByType {
  full_time: number;
  part_time: number;
  contract: number;
  internship: number;
  remote: number;
}

export interface AnalyticsSummary {
  total_jobs: number;
  active_jobs: number;
  total_applications: number;
  total_views: number;
  by_status: JobCountByStatus;
  by_type: JobCountByType;
  top_jobs: JobListItem[];
}

// ── UI helpers ───────────────────────────────────────────────────────────────

export const JOB_TYPE_LABELS: Record<JobType, string> = {
  full_time: "Full Time",
  part_time: "Part Time",
  contract: "Contract",
  internship: "Internship",
  remote: "Remote",
};

export const JOB_STATUS_LABELS: Record<JobStatus, string> = {
  draft: "Draft",
  active: "Active",
  paused: "Paused",
  closed: "Closed",
};

export const JOB_STATUS_COLORS: Record<JobStatus, string> = {
  draft: "bg-gray-100 text-gray-700",
  active: "bg-green-100 text-green-700",
  paused: "bg-yellow-100 text-yellow-700",
  closed: "bg-red-100 text-red-700",
};

File: backend/app/schemas/job.py
"""Pydantic v2 schemas for the job posting module."""
import uuid
from datetime import datetime
from typing import Annotated

from pydantic import BaseModel, ConfigDict, Field, field_validator

from app.models.job import JobStatus, JobType, QuestionType


# ─── Shared base ──────────────────────────────────────────────────────────────
class _Base(BaseModel):
    model_config = ConfigDict(from_attributes=True, populate_by_name=True)


# ─── Skills ───────────────────────────────────────────────────────────────────
class JobSkillCreate(_Base):
    skill_name: Annotated[str, Field(min_length=1, max_length=100)]
    is_required: bool = True


class JobSkillResponse(_Base):
    id: uuid.UUID
    skill_name: str
    is_required: bool


# ─── Questions ────────────────────────────────────────────────────────────────
class JobQuestionCreate(_Base):
    question_text: Annotated[str, Field(min_length=1)]
    question_type: QuestionType = QuestionType.TEXT
    options: list[str] | None = None
    is_required: bool = True
    display_order: int = 0

    @field_validator("options")
    @classmethod
    def validate_options(cls, v: list[str] | None, info: object) -> list[str] | None:
        # options only relevant for multiple_choice
        if v is not None and len(v) < 2:
            raise ValueError("multiple_choice questions need at least 2 options")
        return v


class JobQuestionUpdate(_Base):
    question_text: str | None = None
    question_type: QuestionType | None = None
    options: list[str] | None = None
    is_required: bool | None = None
    display_order: int | None = None


class JobQuestionResponse(_Base):
    id: uuid.UUID
    question_text: str
    question_type: QuestionType
    options: list[str] | None
    is_required: bool
    display_order: int


class QuestionsReorderRequest(_Base):
    """List of question IDs in their new order (first = order 0)."""
    question_ids: list[uuid.UUID]


# ─── Pipeline stages ──────────────────────────────────────────────────────────
class PipelineStageCreate(_Base):
    stage_name: Annotated[str, Field(min_length=1, max_length=100)]
    color: str = "#6366f1"


class PipelineStageUpdate(_Base):
    stage_name: str | None = None
    color: str | None = None


class PipelineStageReorderItem(_Base):
    id: uuid.UUID
    stage_order: int


class PipelineStageResponse(_Base):
    id: uuid.UUID
    stage_name: str
    stage_order: int
    color: str
    is_default: bool


# ─── Job CRUD ─────────────────────────────────────────────────────────────────
class JobCreate(_Base):
    title: Annotated[str, Field(min_length=1, max_length=255)]
    description: str | None = None
    requirements: str | None = None
    location: str | None = None
    job_type: JobType = JobType.FULL_TIME
    department: str | None = None
    salary_min: int | None = None
    salary_max: int | None = None
    currency: Annotated[str, Field(min_length=3, max_length=3)] = "USD"
    experience_years_min: int | None = None
    experience_years_max: int | None = None
    deadline: datetime | None = None

    @field_validator("salary_max")
    @classmethod
    def salary_range_valid(cls, v: int | None, info: object) -> int | None:
        if v is not None:
            data = getattr(info, "data", {})
            sal_min = data.get("salary_min")
            if sal_min is not None and v < sal_min:
                raise ValueError("salary_max must be >= salary_min")
        return v


class JobUpdate(_Base):
    title: Annotated[str, Field(min_length=1, max_length=255)] | None = None
    description: str | None = None
    requirements: str | None = None
    location: str | None = None
    job_type: JobType | None = None
    department: str | None = None
    salary_min: int | None = None
    salary_max: int | None = None
    currency: Annotated[str, Field(min_length=3, max_length=3)] | None = None
    experience_years_min: int | None = None
    experience_years_max: int | None = None
    deadline: datetime | None = None


class JobStatusUpdateRequest(_Base):
    status: JobStatus


# ─── Job responses ────────────────────────────────────────────────────────────
class JobResponse(_Base):
    id: uuid.UUID
    title: str
    description: str | None
    requirements: str | None
    location: str | None
    job_type: JobType
    department: str | None
    salary_min: int | None
    salary_max: int | None
    currency: str
    experience_years_min: int | None
    experience_years_max: int | None
    status: JobStatus
    posted_by: uuid.UUID | None
    company_id: uuid.UUID | None
    published_at: datetime | None
    closed_at: datetime | None
    archived_at: datetime | None
    deadline: datetime | None
    views_count: int
    applications_count: int
    created_at: datetime
    updated_at: datetime
    skills: list[JobSkillResponse]
    questions: list[JobQuestionResponse]
    pipeline_stages: list[PipelineStageResponse]


class JobListItem(_Base):
    """Lightweight projection used in list endpoints."""
    id: uuid.UUID
    title: str
    location: str | None
    job_type: JobType
    department: str | None
    status: JobStatus
    views_count: int
    applications_count: int
    created_at: datetime
    published_at: datetime | None
    deadline: datetime | None
    skills: list[JobSkillResponse] = []


class JobListResponse(_Base):
    items: list[JobListItem]
    total: int
    page: int
    page_size: int


# ─── Analytics ────────────────────────────────────────────────────────────────
class JobCountByStatus(_Base):
    draft: int = 0
    active: int = 0
    paused: int = 0
    closed: int = 0


class JobCountByType(_Base):
    full_time: int = 0
    part_time: int = 0
    contract: int = 0
    internship: int = 0
    remote: int = 0


class AnalyticsSummary(_Base):
    total_jobs: int
    active_jobs: int
    total_applications: int
    total_views: int
    by_status: JobCountByStatus
    by_type: JobCountByType
    top_jobs: list[JobListItem]
